fuse_constraints/src/fixed_3d_landmark_simple_covariance_constraint.cpp

```diff
@@ -0,0 +1,116 @@
+#include <pluginlib/class_list_macros.hpp>
+
+#include <boost/serialization/export.hpp>
+#include <ceres/autodiff_cost_function.h>
+#include <Eigen/Dense>
+
+#include <string>
+
+namespace fuse_constraints
+{
+
+Fixed3DLandmarkSimpleCovarianceConstraint::Fixed3DLandmarkSimpleCovarianceConstraint(
+    const std::string& source, const fuse_variables::Position3DStamped& position,
+    const fuse_variables::Orientation3DStamped& orientation, const fuse_variables::PinholeCamera& calibration,
+    const fuse_core::MatrixXd& pts3d, const fuse_core::MatrixXd& observations, const fuse_core::Vector7d& mean,
+    const fuse_core::Matrix2d& covariance)
+  : fuse_core::Constraint(source, { position.uuid(), orientation.uuid(), calibration.uuid() })
+  , pts3d_(pts3d)
+  , observations_(observations)
+  , mean_(mean)
+  , sqrt_information_(covariance.inverse().llt().matrixU())
+{
+  assert(pts3d_.cols() == 3);
+  assert(observations_.cols() == 2);
+  assert(pts3d_.rows() == observations_.rows());
+}
+
+Fixed3DLandmarkSimpleCovarianceConstraint::Fixed3DLandmarkSimpleCovarianceConstraint(
+    const std::string& source, const fuse_variables::Position3DStamped& position,
+    const fuse_variables::Orientation3DStamped& orientation, const fuse_variables::PinholeCamera& calibration,
+    const double& marker_size, const fuse_core::MatrixXd& observations, const fuse_core::Vector7d& mean,
+    const fuse_core::Matrix2d& covariance)
+  : fuse_core::Constraint(source, { position.uuid(), orientation.uuid(), calibration.uuid() })
+  , pts3d_(4, 3)
+  , observations_(observations)
+  , mean_(mean)
+  , sqrt_information_(covariance.inverse().llt().matrixU())
+{
+  // Define 3D Homogeneous 3D Points at origin, assume z-up
+  pts3d_ << -1.0, -1.0, 0.0,  // NOLINT
+            -1.0,  1.0, 0.0,  // NOLINT
+             1.0, -1.0, 0.0,  // NOLINT
+             1.0,  1.0, 0.0;  // NOLINT
+  pts3d_ *= marker_size;      // Scalar Multiplication
+
+  assert(pts3d_.cols() == 3);
+  assert(observations_.cols() == 2);
+  assert(pts3d_.rows() == observations_.rows());
+}
+
+void Fixed3DLandmarkSimpleCovarianceConstraint::print(std::ostream& stream) const
+{
+  stream << type() << "\n"
+         << "  source: " << source() << "\n"
+         << "  uuid: " << uuid() << "\n"
+         << "  position variable: " << variables().at(0) << "\n"
+         << "  orientation variable: " << variables().at(1) << "\n"
+         << "  mean: " << mean().transpose() << "\n"
+         << "  sqrt_info: " << sqrtInformation() << "\n"
+         << "  observations: " << observations() << "\n";
+
+  if (loss())
+  {
+    stream << "  loss: ";
+    loss()->print(stream);
+  }
+}
+
+ceres::CostFunction* Fixed3DLandmarkSimpleCovarianceConstraint::costFunction() const
+{
+  return new ceres::AutoDiffCostFunction<Fixed3DLandmarkSimpleCovarianceCostFunctor, 8, 3, 4, 4>(
+      new Fixed3DLandmarkSimpleCovarianceCostFunctor(sqrt_information_, mean_, observations_, pts3d_));
+}
```

Comment:
A question about the numbers here, 8, 3, 4, 4>? The first parameter is the dimension of the residues. If we look into the implementation of the CostFunctor inside  ' Fixed3DLandmarkSimpleCovarianceCostFunctor', it will generate 2*N residules, where N is the number of observations. Should it be something dynamic or it doesn't matter?